_rules/table-header-cells-refer-to-data-cells-d0f69e.md

```diff
@@ -0,0 +1,246 @@
+---
+id: d0f69e
+name: All table header cells have assigned data cells
+rule_type: atomic
+description: |
+  This rule checks that each table header has assigned data cells in a table element.
+accessibility_requirements:
+  wcag20:1.3.1: # Info and Relationships (A)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+input_aspects:
+  - DOM Tree
+acknowledgements:
+  authors:
+    - Jey Nandakumar
+    - Audrey Maniez
+---
+
+## Applicability
+
+The rule applies to any HTML element that is a [descendant](https://dom.spec.whatwg.org/#concept-tree-descendant) in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) of an element having a [semantic role](#semantic-role) of either [table](https://www.w3.org/TR/wai-aria-1.1/#table) or [grid](https://www.w3.org/TR/wai-aria-1.1/#grid), that is [visible](#visible) as well as [included in the accessibility tree](#included-in-the-accessibility-tree), where the element has any of the following [semantic roles](#semantic-role):
+
+- [rowheader](https://www.w3.org/TR/wai-aria-1.1/#rowheader)
+- [columnheader](https://www.w3.org/TR/wai-aria-1.1/#columnheader)
+
+## Expectation
+
+Each target element is [assigned](https://html.spec.whatwg.org/multipage/tables.html#header-and-data-cell-semantics) to at least one element with a [semantic role](#semantic-role) of [cell](https://www.w3.org/TR/wai-aria-1.1/#cell) or [gridcell](https://www.w3.org/TR/wai-aria-1.1/#gridcell), that is a [descendant](https://dom.spec.whatwg.org/#concept-tree-descendant) in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) of the same [table](https://www.w3.org/TR/wai-aria-1.1/#table) or [grid](https://www.w3.org/TR/wai-aria-1.1/#grid) element, as that of the target element.
```

Comment:
I'm still not sure that the "descendant of the same table" part is not correct in case of nested tables (where cells in the inner table are descendant of both the inner and outer `table` element).
I think it's OK because the assignation algorithm can only assign to cell "in the same table".

Maybe it's worth changing this to "…that is a cell in the same table as the target element". Or adding a note to make that point explicit.

Reply:
I think it is OK as is, because the assignation algorithm as you point out can only point to cell in the same table.

Reply:
Actually, if we rely on the assignation algorithm for pointing into the same table, this can be completely shortened to "each target element is assigned to at least one cell or gridcell" (i.e., no need for the "that is in the same table" part). + a note pointing out that this will necessarily be in the same table.

Reply:
Totally agreed. Although keep in mind that the header assignment algorithm does not cross shadow boundaries.